### PR TITLE
[Fix #10204] Add new Lint/GlobalExceptionVars cop

### LINKED ISSUE #10204: Introduce cop to warn against usage of `$!` and `$@`.
https://bugs.ruby-lang.org/issues/18083#note-21

Introduce a cop to warn against all usage of `$!` and `$@`. There is no safe usage except for perhaps `expression rescue $!`.

Usage of `$!` can be rewritten:

```ruby

begin
rescue => error
  raise
ensure
  if error ...
end
```

But Matz also said he disagree with such an approach. But at least it's more safe than `if $!`.

### PR DESCRIPTION
**Pull request by andyw8**
> [!NOTE]
> This PR was co-authored with [Claude Code](https://claude.com/claude-code).

This PR adds a new `Lint/ImplicitExceptionVars` cop to warn against usage of implicit exception variables as discussed in #10204.

Implementation is based on @ioquatix's [GlobalExceptionVariables cop](https://github.com/socketry/rubocop-socketry/commit/025f7c3d78b1f33b73eaf3a39bc071e3d7747f0e).

## Summary

The cop detects usage of:
- `$!` (last exception)
- `$@` (backtrace of last exception)  
- `$ERROR_INFO` (English alias for `$!`)
- `$ERROR_POSITION` (English alias for `$@`)

These variables are unsafe because they reference the last exception raised, which can change unexpectedly in complex code. Using explicit exception variables in rescue blocks is clearer and safer.

## Example

```ruby
# bad
begin
  do_something
rescue
  puts $!
  puts $@
end

# good
begin
  do_something
rescue => e
  puts e
  puts e.backtrace
end
```

## Implementation Details

- **Department**: `Lint` (as discussed in the issue)
- **Enabled**: `pending` (users must explicitly enable it)
- **Autocorrect**: No (requires manual refactoring)
- **Tests**: Comprehensive test coverage included

## Interaction with Style/SpecialGlobalVars

Both cops can fire together when `$!` or `$@` is used, which is intentional:
- `Style/SpecialGlobalVars`: Style preference (if you use these, use English names)
- `Lint/ImplicitExceptionVars`: Safety warning (don't use these at all)

Since the new cop is pending by default, it won't interfere with existing projects unless explicitly enabled.

## Checklist

- [x] The PR relates to only one subject with a clear title and description
- [x] Wrote good commit messages
- [x] Commit message starts with `[Fix #10204]`
- [x] Feature branch is up-to-date with `master`
- [x] Added tests
- [x] Ran `bundle exec rake default` - all tests pass ✅
- [x] Added changelog entry

### CODE REVIEW

File: lib/rubocop/result_cache.rb [hunk 1]
```diff
@@ -58,10 +58,10 @@ def remove_oldest_files(files, rubocop_cache_dir, verbose)
         puts "Removing the #{remove_count} oldest files from #{rubocop_cache_dir}" if verbose
         sorted = files.sort_by { |path| File.mtime(path) }
         remove_files(sorted, remove_count)
-      rescue Errno::ENOENT
+      rescue Errno::ENOENT => e
         # This can happen if parallel RuboCop invocations try to remove the
         # same files. No problem.
-        puts $ERROR_INFO if verbose
```
Comment (andyw8) [inline lib/rubocop/result_cache.rb:64]: Only one violation in RuboCop!

Comment (ioquatix) [reply lib/rubocop/result_cache.rb:64]: This is actually an improvement too!

File: spec/rubocop/cop/lint/implicit_exception_vars_spec.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    RUBY
+  end
+
+  it 'registers an offense for $@' do
+    expect_offense(<<~RUBY)
+      begin
+        do_something
+      rescue
+        puts $@
+             ^^ Avoid implicit exception variables `$@`. Use explicit exception variable in rescue instead.
```
Comment (koic) [inline spec/rubocop/cop/lint/implicit_exception_vars_spec.rb]: Considering user convenience, it would be necessary to explicitly state in the message that backtrace is used.

File: config/default.yml [hunk 3]
```diff
@@ -2021,6 +2021,11 @@ Lint/IdentityComparison:
   StyleGuide: '#identity-comparison'
   VersionAdded: '0.91'
 
+Lint/ImplicitExceptionVars:
+  Description: 'Checks for uses of implicit exception variables like `$!` and `$@`.'
+  Enabled: pending
+  VersionAdded: '<<next>>'
+
```
Comment (ydakuka) [inline config/default.yml]: > This PR was co-authored with [Claude Code](https://claude.com/claude-code).

I think implementing (or updating) a RuboCop cop in a single iteration with only one AI agent is not sufficient.

I usually ask "Gemini Pro" like this:
```
module RuboCop
  module Cop
    module Lint
      # Checks for uses of implicit exception variables like `$!` and `$@`.
      # ...
    end
  end
end

RSpec.describe RuboCop::Cop::Lint::ImplicitExceptionVars, :config do
  it 'registers an offense for $!' do
  # ...

  it 'does not register an offense for other global variables' do
    expect_no_offenses(<<~RUBY)
      puts $stdin
      puts $stdout
      puts $$
    RUBY
  end
end

Please analyze this RuboCop cop in detail and check whether all edge and corner cases are covered.
```

Comment (ydakuka) [reply config/default.yml]: Then I paste Gemini Pro’s answer into Claude Code so it can be fixed.

Comment (ydakuka) [reply config/default.yml]: Several interactions like this, along with manual review, provide pretty good coverage of the cop.

Comment (ydakuka) [reply config/default.yml]: > A Gemini Pro answer for this cop:

This is a well-structured and idiomatic RuboCop cop. It correctly inherits from `Base`, uses idiomatic node matching (`on_gvar`), and defines clear error messages.

However, regarding your question about whether **all** edge and corner cases are covered: **No, there are specific Ruby edge cases regarding global variables that this implementation misses.**

Below is a detailed analysis of the coverage gaps and how to fix them.

### 1. Missing Coverage: Global Variable Assignment (`gvasgn`)

The current implementation only defines `on_gvar`. In RuboCop's AST (Abstract Syntax Tree), `on_gvar` handles **referencing** (reading) a global variable. It does **not** handle **assigning** to one.

While assigning to `$!` is rare and generally discouraged, it is valid Ruby (often used in legacy code to "clear" the current exception).

**The Gap:**

```ruby
# This will NOT be caught by the current cop
begin
  do_something
rescue
  $! = nil  # Clears the exception
  $ERROR_INFO = nil
end

```

**The Fix:**
You need to implement `on_gvasgn` (Global Variable Assignment) to catch these cases.

```ruby
def on_gvasgn(node)
  return unless IMPLICIT_VARS.include?(node.name)

  add_offense(node, message: format(MSG, var: node.name))
end

```

### 2. Edge Case: `alias` keyword

Ruby allows aliasing global variables.

```ruby
alias $my_custom_error $!

```

**Current Behavior:**

1. The line `alias $my_custom_error $!` **will** be flagged (because `$!` appears as a `gvar` node in the alias definition).
2. However, subsequent usage of `$my_custom_error` will **not** be flagged.

**Verdict:** This is likely an acceptable limitation. Tracking aliases requires complex state tracking across the codebase, which is usually out of scope for a single Lint cop. However, it is worth noting that a user could bypass this check via aliasing.

### 3. Edge Case: `defined?` keyword

The `defined?` keyword is a special form in Ruby.

```ruby
if defined?($!)
  puts "We have a problem"
end

```

**Current Behavior:**
RuboCop's traversal generally descends into `defined?` nodes. Inside `defined?($!)`, the `$!` is parsed as a `gvar` node.

* **Result:** This **will** be caught by `on_gvar`.
* **Is this correct?** Yes. Relying on `defined?($!)` is just as unsafe/implicit as reading `$!`.

### 4. Missing Feature: Autocorrection

While not strictly a "coverage" issue regarding detection, the cop does not support `AutoCorrect`.

**Why this is hard:**
Autocorrecting this is non-trivial because it requires context awareness:

1. **If inside a rescue:** It must check if an explicit variable already exists.
* *Scenario A:* `rescue => e; puts $!; end` -> `rescue => e; puts e; end`
* *Scenario B:* `rescue; puts $!; end` -> `rescue => e; puts e; end` (Requires generating a unique variable name to avoid shadowing).


2. **If outside a rescue:** (e.g., in a helper method), it is impossible to autocorrect safely because there is no `rescue` node to attach the variable to.

**Recommendation:** It is acceptable that this cop does not autocorrect, as manual intervention is usually required to refactor the logic safely.

### 5. Suggested Improvements

Here is the refined code including the `gvasgn` fix and an expanded test suite to prove the edge cases.

#### Updated Cop Code

```ruby
# frozen_string_literal: true

module RuboCop
  module Cop
    module Lint
      class ImplicitExceptionVars < Base
        MSG = 'Avoid implicit exception variables `%<var>s`. ' \
              'Use explicit exception variable in rescue instead.'

        IMPLICIT_VARS = %i[$! $@ $ERROR_INFO $ERROR_POSITION].freeze

        def on_gvar(node)
          check_implicit_var(node)
        end

        # Added handler for assignment
        def on_gvasgn(node)
          check_implicit_var(node)
        end

        private

        def check_implicit_var(node)
          return unless IMPLICIT_VARS.include?(node.name)

          add_offense(node, message: format(MSG, var: node.name))
        end
      end
    end
  end
end

```

#### Additional Tests to Cover Edge Cases

You should add these tests to your RSpec file to ensure robustness:

```ruby
  it 'registers an offense when assigning to $!' do
    expect_offense(<<~RUBY)
      begin
        do_something
      rescue
        $! = nil
        ^^ Avoid implicit exception variables `$!`. Use explicit exception variable in rescue instead.
      end
    RUBY
  end

  it 'registers an offense when assigning to $ERROR_INFO' do
    expect_offense(<<~RUBY)
      require 'English'
      $ERROR_INFO = nil
      ^^^^^^^^^^^ Avoid implicit exception variables `$ERROR_INFO`. Use explicit exception variable in rescue instead.
    RUBY
  end

  it 'registers an offense when used inside defined?' do
    expect_offense(<<~RUBY)
      if defined?($!)
                  ^^ Avoid implicit exception variables `$!`. Use explicit exception variable in rescue instead.
        puts 'error present'
      end
    RUBY
  end

  it 'registers an offense when aliasing $!' do
    expect_offense(<<~RUBY)
      alias $my_error $!
                      ^^ Avoid implicit exception variables `$!`. Use explicit exception variable in rescue instead.
    RUBY
  end

```

### Summary

The original code was 90% there. It correctly identified standard usage but missed **assignments** to these global variables. Adding `on_gvasgn` ensures that even obscure manipulations of the global exception state are flagged.

Comment (koic) [reply config/default.yml]: Different workflows work for different people, but relying solely on AI-generated output is often not sufficient for cop changes. AI can help explore cases, but a careful human self-review and follow-up testing are what ultimately bring the work to PR-ready quality.

Comment (bbatsov) [reply config/default.yml]: Agreed. I'm all for using AI tooling, but in the end of the day it's important to verify the proposed code changes with some old-school real thinking. 😁 

@ydakuka Good tips on how to use different agents on the same task, though. You might want to write some article on the subject.

**Comment by andyw8** [comment]
Thanks for the feedback so far, I've made some updates.

The error text needs some adjustment still, as it doesn't make sense in all contexts (see the tests). Further feedback welcome.

Edit: ✅ error text updated

File: lib/rubocop/cop/lint/implicit_exception_vars.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      #   end
+      #
+      #   # good
+      #   begin
+      #     do_something
+      #   rescue StandardError => error
+      #     puts error
+      #     puts error.backtrace
+      #   end
+      class ImplicitExceptionVars < Base
```
Comment (koic) [inline lib/rubocop/cop/lint/implicit_exception_vars.rb]: I'm concerned that the name `Lint/ImplicitExceptionVars` could be misleading. The following example implicitly relies on the current exception without introducing an explicit exception variable:

```ruby
begin
  raise_exception
rescue
  raise
end
```

For that reason, it seems preferable to choose a more appropriate cop name. It might be better if the name more clearly reflects that this cop targets implicit global exception variables.

**Comment by andyw8** [comment]
I've renamed it to "GlobalExceptionVars". I don't think prefixing it with "Implicit" adds anything useful.

I went with `Vars` instead of `Variables` to follow the existing naming patterns

**Comment by andyw8** [comment]
:wave: Any further feedback on this? I believe everything mentioned so far has been addressed.

**Comment by ioquatix** [comment]
@koic worth moving forward with this?

**Comment by andyw8** [comment]
I'm happy to address any remaining concerns.
